app/api/dds/v1/search_api.rb

```diff
@@ -76,12 +78,81 @@ class SearchAPI < Grape::API
+          optional :query, type: String, desc: "The string fragment (phrase) to search for in query_string.fields. Required if query_string.fields are submitted."
+          optional :fields, type: Array, desc: "List of text fields to search; valid options are name and tags.label.  If not specified, defaults to all options. Requires query_string.query."
+        end
+        optional :filters, type: Array, desc: 'List of boolean search objects (predicates) that are joined together by AND operator, represented as key: value pairs' do
+          optional :kind , type: Array, desc: 'Limit search context to list of kinds; valid options are dds-folder and dds-file - requires a set of comma-delimited kinds like so: {"kind": ["dds-file"]}. If not specified, defaults to all options.'
+          optional 'project.id', type: Array, desc: 'Limit search context to list of projects; requires a set of comma-delimited ids like so: {"project.id": ["345e...", "2e45..."]}. Filters out projects to which the user is not allowed access. If not specified, defaults to projects in which the current user has view permission.'
+        end
+        optional :aggs, type: Array, desc: 'List of aggs (aggregates) that should be computed.' do
+          optional :field, type: String, desc: 'The field for which you want to compute agg; valid options are project.name and tags.label. Required if one or more aggs are submitted.'
+          optional :name, type: String, desc: 'The name to give the group of computed agg buckets for the field. Required if one or more aggs are submitted.'
+          optional :size, type: Integer, desc: 'The number of agg buckets to return with the results; defaults to the top 20 and cannot be greater than 50.'
+        end
+        optional :post_filters, type: Array, desc: 'List of boolean search objects (predicates) that are joined together by the AND operator - they are represented as key: value pairs' do
+          optional 'project.name', type: Array, desc: 'List of project names; requires a set of comma-delimited names like so: {"project.name": ["Big Mouse", "BD2K"]}.'
+          optional 'tags.label', type: Array, desc: 'List of tags; requires a set of comma-delimited tag labels like so: {"tags.label": ["sequencing", "DNA"]}.'
+        end
+        use :pagination
+      end
+      rescue_from ArgumentError do |e|
+        error_json = {
+          "error" => "400",
+          "reason" => e.message,
+          "suggestion" => "Please supply the correct argument"
+        }
+        error!(error_json, 400)
+      end
+      post '/search/folders_files', root: false do
+        authenticate!
+        search_params = declared(params, include_missing: false)
+        project_filter_exists = false
+
+        search_params[:filters] ||= []
+        search_params[:filters].each do |f|
+          if f.has_key? 'project.id'
+            filtered = policy_scope(Project).where(is_deleted: false, id: f['project.id']).all.to_a.map{|p| p.id}
```

Comment:
There should only ever be at most one 'project.id', or 'kind' entry in the search_params[:filters] array. This will not call the policy_scope more than once per request. I could throw an exception for this in the FoldersFilesResponse.filter or validate_filters method to be sure.